modules/core/src/test/java/org/apache/ignite/internal/processors/cache/datastructures/IgniteReentrantLockTest.java

```diff
@@ -0,0 +1,309 @@
+                        }
+
+                        return null;
+                    }
+
+                    try {
+                        assertTrue(lock.isLocked());
+
+                        Thread.sleep(1_000L);
+                    }
+                    finally {
+                        lock.unlock();
+                    }
+
+                    return null;
+                }
+            }));
+        }
+
+        for (IgniteInternalFuture<?> fut : futs)
+            fut.get(60_000L);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailoverMultilocksUnfair() throws Exception {
+        testReentrantLockMultinodeFailoverMultilocks(false);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailoverMultilocksFair() throws Exception {
+        testReentrantLockMultinodeFailoverMultilocks(true);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinodeFailoverMultilocks(final boolean fair) throws Exception {
+        List<IgniteInternalFuture<Void>> futs = new ArrayList<>(NODES_CNT*2);
+
+        for (int i = 0; i < NODES_CNT; i++) {
+            final int inx = i;
+            final boolean flag = RND.nextBoolean();
+
+            futs.add(GridTestUtils.runAsync(new Callable<Void>() {
+                @Override public Void call() throws Exception {
+                    IgniteLock lock1 = createReentrantLock(inx, fair ? "lock1f": "lock1u", fair);
+                    IgniteLock lock2 = createReentrantLock(inx, fair ? "lock2f": "lock2u", fair);
+
+                    UUID id = grid(inx).cluster().localNode().id();
+
+                    lock1.lock();
+                    lock2.lock();
+
+                    if (flag) {
+                        try {
+                            grid(inx).close();
+                        }
+                        catch (Exception ignored) {
+                            lock2.unlock();
+                            lock1.unlock();
+                        }
+
+                        return null;
+                    }
+
+                    try {
+                        assertTrue(lock2.isLocked());
+                        assertTrue(lock1.isLocked());
+
+                        Thread.sleep(1_000L);
+                    }
+                    finally {
+                        lock2.unlock();
+                        lock1.unlock();
+                    }
+
+                    return null;
+                }
+            }));
+        }
+
+
+        for (IgniteInternalFuture<?> fut : futs)
+            fut.get(60_000L);
+    }
+
+    /**
+     * @param lockName Reentrant lock name.
+     * @param fair Fairness flag.
+     * @return Distributed reentrant lock.
+     * @throws Exception If failed.
+     */
+    private IgniteLock createReentrantLock(int cnt, String lockName, boolean fair) throws Exception {
```

Comment:
may be rename to getOrCreateReentrantLock

Reply:
Another creat lock method has the same name.

Reply:
if somebody writes bad code you will continue write bad code?)

Reply:
I think this is a task for the other issue in Jira:-)